### PR TITLE
feat: memoize build function

### PR DESCRIPTION
**Pull request by eduardoboucas**
**- Summary**

Use memoization instead of debouncing the build function.

**- Test plan**

Reinstates the TypeScript tests.

**- A picture of a cute animal (not mandatory but encouraged)**

cat-sample_1313

### CODE REVIEW

File: src/utils/serve-functions.js [hunk 1]
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
...
+    const key = `${eventType}@${path}`
+
+    if (!cache.has(key)) {
+      cache.set(
+        key,
+        buildFunction(path, eventType).finally(() => cache.delete(key)),
+      )
+    }
+
+    return cache.get(key)
```
Comment (erezrokah) [inline src/utils/serve-functions.js:534]: I couldn't find where the return value is used

File: src/utils/serve-functions.js [hunk 2]
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
...
-    leading: true,
-    trailing: true,
-  })
+  const cache = new Map()
+
+  const buildFunction = getBuildFunction({ functionBuilder, log })
+
+  // `memoizedBuild` will avoid consecutive calls to the build function with
+  // the same arguments until the previous invokation has finished running.
+  const memoizedBuild = (path, eventType) => {
```
Comment (erezrokah) [inline src/utils/serve-functions.js:524]: If the user edits the same file multiple times only the first triggers a build?
Do we want to run the last one too?

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @eduardoboucas, I think we might need a combination of debounce and memoize

File: src/utils/serve-functions.js (hunk 1 above)
Comment (eduardoboucas) [reply src/utils/serve-functions.js:534]: It's not used currently, but `debouncedBuild` preserved the return value of the build function and I thought `memoizedBuild` should respect that.

File: src/utils/serve-functions.js [hunk 3]
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
...
+    const key = `${eventType}@${path}`
+
+    if (!cache.has(key)) {
+      cache.set(
+        key,
+        buildFunction(path, eventType).finally(() => cache.delete(key)),
```
Comment (ehmicky) [inline src/utils/serve-functions.js:530]: Would using an async-aware memoization library like [`moize`](https://github.com/planttheidea/moize) work here?
https://github.com/planttheidea/moize#ispromise

**Comment by eduardoboucas** [comment]
In https://github.com/netlify/cli/pull/2258/commits/b9f99f49130755c5d431bda31a3a00a16b10f1de I've updated the `memoizeBuild` to ensure any updates performed while the function builder is running are executed afterwards, whilst at the same time being able to discard any duplicate filesystem events in the same way that `debounce` used to do.

The implementation is non-trivial, but hopefully the comments I've added help clarify the behaviour.

> If the user edits the same file multiple times only the first triggers a build?
> Do we want to run the last one too?

@erezrokah This should now be addressed.

> Would using an async-aware memoization library like moize work here?

I think this is a special case where we want a time-based memoization with a debounce interval. I'm not aware of any library that does exactly what we need, but I'm happy to explore alternatives.

**Review by erezrokah — APPROVED** [review]

**Review by ehmicky — APPROVED** [review]

**Review by erezrokah — APPROVED** [review]

**Pull request merged**